File: platform/macos/Sources/HandShakerCore/Models/Event.swift
import Foundation

/// Backend event kinds (event.rs `BackendEvent`). Rust serializes with
/// `#[serde(tag = "kind", rename_all = "snake_case")]`: every variant's
/// fields are *inlined* next to the `kind` tag (verified by
/// `backend_event_change_payloads_serialize_with_stable_kinds` in
/// tests.rs), e.g.:
///
///     {"kind":"transfer_updated","id":7,"session_id":1,...}
///     {"kind":"warning","code":"...","message":"...","detail":null,...}
///
/// Unknown `kind` tokens decode safely as `.unknown(String)` instead of
/// failing, so a newer Rust library cannot break the Swift client.
public enum BackendEvent: Codable, Sendable, Equatable {
    case runtimeStarted
    case runtimeStopping
    /// Newtype payload inlined: the DeviceDescriptor fields sit next to
    /// "kind":"device_added".
    case deviceAdded(DeviceDescriptor)
    case deviceUpdated(sessionID: UInt64, device: DeviceDescriptor)
    case deviceRemoved(deviceID: DeviceID)
    /// Newtype payload inlined: SessionSnapshot fields.
    case sessionStateChanged(SessionSnapshot)
    /// Newtype payload inlined: TransferSnapshot fields.
    case transferUpdated(TransferSnapshot)
    case connectionLost(sessionID: UInt64)
    case clipboardChanged(sessionID: UInt64, entries: [ClipboardEntry])
    case mediaChanged(sessionID: UInt64, change: MediaChange)
    case remoteFileChanged(sessionID: UInt64, change: RemoteFileChange)
    /// Newtype payload inlined: SyncRunResult fields.
    case syncWatchApplied(SyncRunResult)
    /// Newtype payload inlined: PublicError fields.
    case warning(HandShakerNativeError)
    /// Forward compatibility: raw `kind` token of an unknown event.
    case unknown(String)

    private enum CodingKeys: String, CodingKey {
        case kind
        case sessionID = "session_id"
        case deviceID = "device_id"
        case entries
        case change
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let kind = try container.decode(String.self, forKey: .kind)
        switch kind {
        case "runtime_started":
            self = .runtimeStarted
        case "runtime_stopping":
            self = .runtimeStopping
        case "device_added":
            self = .deviceAdded(try DeviceDescriptor(from: decoder))
        case "device_updated":
            let sessionID = try container.decode(UInt64.self, forKey: .sessionID)
            self = .deviceUpdated(sessionID: sessionID, device: try DeviceDescriptor(from: decoder))
        case "device_removed":
            self = .deviceRemoved(deviceID: try container.decode(DeviceID.self, forKey: .deviceID))
        case "session_state_changed":
            self = .sessionStateChanged(try SessionSnapshot(from: decoder))
        case "transfer_updated":
            self = .transferUpdated(try TransferSnapshot(from: decoder))
        case "connection_lost":
            self = .connectionLost(sessionID: try container.decode(UInt64.self, forKey: .sessionID))
        case "clipboard_changed":
            self = .clipboardChanged(
                sessionID: try container.decode(UInt64.self, forKey: .sessionID),
                entries: try container.decode([ClipboardEntry].self, forKey: .entries)
            )
        case "media_changed":
            self = .mediaChanged(
                sessionID: try container.decode(UInt64.self, forKey: .sessionID),
                change: try container.decode(MediaChange.self, forKey: .change)
            )
        case "remote_file_changed":
            self = .remoteFileChanged(
                sessionID: try container.decode(UInt64.self, forKey: .sessionID),
                change: try container.decode(RemoteFileChange.self, forKey: .change)
            )
        case "sync_watch_applied":
            self = .syncWatchApplied(try SyncRunResult(from: decoder))
        case "warning":
            self = .warning(try HandShakerNativeError(from: decoder))
        default:
            self = .unknown(kind)
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .runtimeStarted:
            try container.encode("runtime_started", forKey: .kind)
        case .runtimeStopping:
            try container.encode("runtime_stopping", forKey: .kind)
        case .deviceAdded(let device):
            try container.encode("device_added", forKey: .kind)
            try device.encode(to: encoder)
        case .deviceUpdated(let sessionID, let device):
            try container.encode("device_updated", forKey: .kind)
            try container.encode(sessionID, forKey: .sessionID)
            try device.encode(to: encoder)
        case .deviceRemoved(let deviceID):
            try container.encode("device_removed", forKey: .kind)
            try container.encode(deviceID, forKey: .deviceID)
        case .sessionStateChanged(let snapshot):
            try container.encode("session_state_changed", forKey: .kind)
            try snapshot.encode(to: encoder)
        case .transferUpdated(let snapshot):
            try container.encode("transfer_updated", forKey: .kind)
            try snapshot.encode(to: encoder)
        case .connectionLost(let sessionID):
            try container.encode("connection_lost", forKey: .kind)
            try container.encode(sessionID, forKey: .sessionID)
        case .clipboardChanged(let sessionID, let entries):
            try container.encode("clipboard_changed", forKey: .kind)
            try container.encode(sessionID, forKey: .sessionID)
            try container.encode(entries, forKey: .entries)
        case .mediaChanged(let sessionID, let change):
            try container.encode("media_changed", forKey: .kind)
            try container.encode(sessionID, forKey: .sessionID)
            try container.encode(change, forKey: .change)
        case .remoteFileChanged(let sessionID, let change):
            try container.encode("remote_file_changed", forKey: .kind)
            try container.encode(sessionID, forKey: .sessionID)
            try container.encode(change, forKey: .change)
        case .syncWatchApplied(let result):
            try container.encode("sync_watch_applied", forKey: .kind)
            try result.encode(to: encoder)
        case .warning(let error):
            try container.encode("warning", forKey: .kind)
            try error.encode(to: encoder)
        case .unknown(let kind):
            try container.encode(kind, forKey: .kind)
        }
    }
}

/// One delivered event with monotonic sequencing (event.rs `EventEnvelope`).
public struct EventEnvelope: Codable, Sendable, Equatable {
    public let sequence: UInt64
    public let timestampMs: UInt64
    public let event: BackendEvent

    private enum CodingKeys: String, CodingKey {
        case sequence
        case timestampMs = "timestamp_ms"
        case event
    }
}

/// Category of a phone-initiated remote file change
/// (dto.rs `RemoteFileChangeKind`, snake_case).
public enum RemoteFileChangeKind: String, Codable, Sendable, Equatable {
    /// A directory monitor event.
    case directoryChanged = "directory_changed"
    /// A synchronization file change.
    case fileChanged = "file_changed"
    /// A one-shot photo synchronization response.
    case photoSyncChanged = "photo_sync_changed"
    /// A synchronization monitor response.
    case syncMonitorChanged = "sync_monitor_changed"
    /// Forward compatibility: unknown kind tokens decode safely.
    case unknown

    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = Self(rawValue: raw) ?? .unknown
    }
}

/// Summarized remote file change (dto.rs `RemoteFileChangeDto`).
/// `files`/`statuses` carry `#[serde(default, skip_serializing_if =
/// "Vec::is_empty")]` in Rust and may be absent from JSON — Swift
/// optionals tolerate that.
public struct RemoteFileChange: Codable, Sendable, Equatable {
    public let changeKind: RemoteFileChangeKind
    public let paths: [String]
    /// Full metadata for each changed path, parallel to `paths` when the
    /// phone supplied it (absent/empty when only paths are known).
    public let files: [FileEntry]?
    /// Per-path `FileChangeStatus` snake_case strings (e.g. "added",
    /// "deleted", "modified"), parallel to `paths`.
    public let statuses: [String]?

    private enum CodingKeys: String, CodingKey {
        case changeKind = "change_kind"
        case paths
        case files
        case statuses
    }
}

File: crates/handshaker-application/src/event.rs
//! Backend event hub (M8 §5.7): a broadcast of `EventEnvelope` with a
//! monotonic sequence per Runtime. Subscribers may lag (broadcast `Lagged`
//! surfaced verbatim); closed subscriptions are identifiable; unknown core
//! events never panic the hub.

use std::sync::Arc;
use std::sync::atomic::{AtomicU64, Ordering};

use tokio::sync::broadcast;

use crate::dto::{
    ClipboardEntryDto, DeviceDescriptor, DeviceId, MediaChangeDto, RemoteFileChangeDto, SessionId,
    SessionSnapshot,
};
use crate::transfer::TransferSnapshot;

/// Backend event kinds for v1 (Session and Transfer are fully implemented;
/// Clipboard/Media/RemoteFile carry phone-initiated change payloads bridged
/// from core (M8.1 Phase C / C1)).
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
#[non_exhaustive]
pub enum BackendEvent {
    RuntimeStarted,
    RuntimeStopping,
    DeviceAdded(DeviceDescriptor),
    DeviceUpdated {
        session_id: SessionId,
        device: DeviceDescriptor,
    },
    DeviceRemoved {
        device_id: DeviceId,
    },
    SessionStateChanged(Box<SessionSnapshot>),
    TransferUpdated(TransferSnapshot),
    /// The core session died under a request or transfer (M8.1 Phase C / C5):
    /// the session is marked `Failed`, its transfers cancelled, and a
    /// `SessionStateChanged(Failed)` event follows.
    ConnectionLost {
        session_id: SessionId,
    },
    /// Clipboard history pushed by the phone.
    ClipboardChanged {
        session_id: SessionId,
        entries: Vec<ClipboardEntryDto>,
    },
    /// A media library change pushed by the phone.
    MediaChanged {
        session_id: SessionId,
        change: MediaChangeDto,
    },
    /// A remote file change (directory monitor / sync) pushed by the phone.
    RemoteFileChanged {
        session_id: SessionId,
        change: RemoteFileChangeDto,
    },
    /// One debounced sync-watch batch was applied incrementally (Phase D /
    /// D6): carries the batch result so GUIs and the CLI can render progress
    /// without polling. `Warning` follows when the batch hit failures.
    /// A sync watch batch was applied incrementally (P1-2: carries
    /// `profile_id` and `session_id` so multi-profile/multi-device UIs can
    /// route the result; the batch result itself is nested under `result`).
    SyncWatchApplied {
        profile_id: String,
        session_id: SessionId,
        result: Box<crate::sync::SyncRunResultDto>,
    },
    Warning(crate::error::PublicError),
}

/// One delivered event with monotonic sequencing.
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct EventEnvelope {
    pub sequence: u64,
    pub timestamp_ms: u64,
    pub event: BackendEvent,
}

fn now_ms() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|duration| duration.as_millis() as u64)
        .unwrap_or(0)
}

/// Broadcast hub owned by a Runtime. `publish` is synchronous (non-blocking
/// send), so progress callbacks can emit events without an async context.
///
/// Closing (M8.1 Phase B / B3): the sender is wrapped in `Option` so
/// `close()` can drop it; every receiver then observes `RecvError::Closed`
/// instead of timing out forever (previously the sender lived as long as the
/// `Arc<RuntimeInner>`, so subscriptions never saw a closed stream).
#[derive(Clone)]
pub struct EventHub {
    tx: Arc<std::sync::Mutex<Option<broadcast::Sender<EventEnvelope>>>>,
    next_sequence: Arc<AtomicU64>,
}

impl EventHub {
    pub fn new(capacity: usize) -> Self {
        let (tx, _) = broadcast::channel(capacity.max(1));
        Self {
            tx: Arc::new(std::sync::Mutex::new(Some(tx))),
            next_sequence: Arc::new(AtomicU64::new(1)),
        }
    }

    /// Publish one event; a full channel drops oldest entries for new
    /// subscribers the same way tokio broadcast does (Lagged on receivers).
    /// Publishing after `close()` is a silent no-op (the hub is gone).
    pub fn publish(&self, event: BackendEvent) {
        let envelope = EventEnvelope {
            sequence: self.next_sequence.fetch_add(1, Ordering::SeqCst),
            timestamp_ms: now_ms(),
            event,
        };
        let guard = self
            .tx
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner());
        if let Some(tx) = guard.as_ref() {
            let _ = tx.send(envelope);
        }
    }

    /// Subscribe with the hub's fixed buffer. `recv()` errors: `Lagged(n)`
    /// when too slow, `Closed` after the hub is closed (runtime shutdown).
    /// Subscribing after close yields a receiver that immediately reports
    /// `Closed`.
    pub fn subscribe(&self) -> broadcast::Receiver<EventEnvelope> {
        let guard = self
            .tx
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner());
        match guard.as_ref() {
            Some(tx) => tx.subscribe(),
            None => {
                let (tx, rx) = broadcast::channel(1);
                drop(tx);
                rx
            }
        }
    }

    /// Close the hub: drops the sender so all receivers observe `Closed`.
    /// Idempotent. After this, `publish` is a no-op and `subscribe` hands
    /// out an already-closed receiver.
    pub fn close(&self) {
        let mut guard = self
            .tx
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner());
        guard.take();
    }

    /// Number of currently active subscribers (diagnostics/tests).
    pub fn subscriber_count(&self) -> usize {
        let guard = self
            .tx
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner());
        guard.as_ref().map(|tx| tx.receiver_count()).unwrap_or(0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::dto::{FileEntryDto, RemoteFileChangeKind};

    fn sample_file(path: &str) -> FileEntryDto {
        FileEntryDto {
            path: path.to_string(),
            size: 7,
            created_at_ms: Some(1000),
            modified_at_ms: Some(2000),
            is_directory: false,
            checksum: Some("abc".to_string()),
            is_trash: Some(false),
            media_id: Some(42),
        }
    }

    #[test]
    fn remote_file_change_v1_payload_stays_byte_identical() {
        // v1 contract: only `change_kind` + `paths` serialize when the
        // optional metadata is absent (fixtures predating files/statuses
        // must keep producing identical JSON).
        let change = RemoteFileChangeDto {
            change_kind: RemoteFileChangeKind::DirectoryChanged,
            paths: vec!["/storage/emulated/0/DCIM".to_string()],
            files: Vec::new(),
            statuses: Vec::new(),
        };
        let value = serde_json::to_value(&change).expect("serialize");
        assert_eq!(value["change_kind"], "directory_changed");
        assert_eq!(value["paths"][0], "/storage/emulated/0/DCIM");
        assert!(value.get("files").is_none(), "empty files must be skipped");
        assert!(
            value.get("statuses").is_none(),
            "empty statuses must be skipped"
        );
    }

    #[test]
    fn remote_file_change_accepts_legacy_json_without_metadata() {
        let value = serde_json::json!({
            "change_kind": "file_changed",
            "paths": ["/storage/emulated/0/a.txt"],
        });
        let change: RemoteFileChangeDto =
            serde_json::from_value(value).expect("legacy payload decodes");
        assert_eq!(change.change_kind, RemoteFileChangeKind::FileChanged);
        assert_eq!(change.paths, vec!["/storage/emulated/0/a.txt"]);
        assert!(change.files.is_empty(), "files default to empty");
        assert!(change.statuses.is_empty(), "statuses default to empty");
    }

    #[test]
    fn remote_file_change_round_trips_full_metadata() {
        let change = RemoteFileChangeDto {
            change_kind: RemoteFileChangeKind::FileChanged,
            paths: vec!["/storage/emulated/0/a.txt".to_string()],
            files: vec![sample_file("/storage/emulated/0/a.txt")],
            statuses: vec!["modified".to_string()],
        };
        let value = serde_json::to_value(&change).expect("serialize");
        assert_eq!(value["files"][0]["path"], "/storage/emulated/0/a.txt");
        assert_eq!(value["files"][0]["size"], 7);
        assert_eq!(value["statuses"][0], "modified");
        let back: RemoteFileChangeDto = serde_json::from_value(value).expect("round trip decodes");
        assert_eq!(back, change);
    }
}
